### PR TITLE
[test] Add opt-in sign-sensitive special-value parity check (copysign, atan2, reciprocal)

### LINKED ISSUE #188680: [test] OpInfo gap: allclose-based compile-vs-eager checks are blind to signed-zero and signed-infinity divergences
# [test] OpInfo gap: allclose-based compile-vs-eager checks are blind to signed-zero and signed-infinity divergences

## Summary

The existing compile-vs-eager correctness check in `TestCommon` uses `torch.allclose`
(or `torch.testing.assert_close`) for output comparison.  IEEE 754 defines
`+0.0 == -0.0`, so `allclose` treats them as identical — meaning a compiler
pass that silently normalises `−0.0 → +0.0` (or flips the sign of an infinity)
will pass every existing correctness test even though it produced the wrong result.

This issue proposes adding a lightweight, opt-in parity check for
`signbit` / `isinf` / `isnan` to the OpInfo test framework so those
divergences can be caught automatically.

---

## Concrete example of the blind spot

```python
import torch

# IEEE 754: +0.0 and -0.0 compare equal
a = torch.tensor(-0.0)
b = torch.tensor(+0.0)

torch.allclose(a, b)           # True  — the sign disappears silently
torch.equal(a, b)              # True  — same
torch.signbit(a).item()        # True  (negative zero)
torch.signbit(b).item()        # False (positive zero)

# The sign of a zero input propagates to the sign of the resulting infinity.
# A pass that normalises -0 → +0 before reciprocal will silently flip the output:
torch.reciprocal(a)            # tensor(-inf)   ← correct
torch.reciprocal(b)            # tensor(inf)    ← wrong, but looks "close"

# For copysign the sign IS the result:
torch.copysign(torch.tensor(0.0), torch.tensor(-1.0))  # tensor(-0.)   eager
# If compiled returned tensor(0.) instead:
torch.allclose(
    torch.tensor(-0.0),
    torch.tensor(+0.0),
)  # True — bug is invisible to the current test
```

If `torch.compile` produced `+0.0` where eager produces `−0.0` for
`torch.copysign`, every existing test would pass. The sign difference would
only be caught by a `signbit`-aware assertion.

---

## Why this matters

Ops whose output sign is semantically load-bearing include (non-exhaustively):

| Op | Sign-sensitive case |
|---|---|
| `torch.copysign` | The sign bit is the entire output |
| `torch.atan2` | IEEE 754 §9.2.1 specifies 4 distinct signed-zero corner cases for boundary inputs |
| `torch.reciprocal` | `1/−0.0 = −inf`, `1/+0.0 = +inf` — sign of zero propagates to sign of inf |
| `torch.neg` | `neg(−0.0) = +0.0` |
| `torch.sign` | `sign(−0.0)` is implementation-defined |
| `torch.pow`, `torch.div` | Various IEEE 754 signed-zero propagation rules |

A compiler optimisation that canonicalises `−0.0 → +0.0` in any of these
will produce silently wrong results that `allclose` cannot detect.

### Related issues

- #53963 — `torch.signbit` itself is currently incorrect for negative zero
  in eager, which means *both* sides of the comparison may be wrong. The
  proposed test infrastructure would surface those cases too.
- #172529 — `torch.isin` eager/compile divergence (shape mismatch), found
  without automated special-value coverage. Demonstrates that compile-vs-eager
  divergences in output structure are already slipping through.
- #164186 — Fuzzer-discovered eager/compile divergence (compile raises,
  eager succeeds). The exception-parity variant of this problem.

---

## Proposed change

Add a single opt-in boolean field to `OpInfo`:

```python
# torch/testing/_internal/opinfo/core.py
check_sign_sensitive_special_values: bool = False
```

And a companion test in `TestCommon` in `test/test_ops.py`:

```python
@ops(op_db, allowed_dtypes=floating_and_complex_types())
def test_sign_sensitive_special_values(self, device, dtype, op):
    """
    Checks that signbit / isinf / isnan are identical between
    eager and torch.compile on the op's reference_inputs.
    """
```

The test:
1. Skips unless `check_sign_sensitive_special_values=True` on the OpInfo
2. Prefers `reference_inputs` (which already generates ±inf, ±0.0 extremal
   combinations) over `sample_inputs`
3. Checks three properties in order:
   - **isnan parity**: NaN positions match
   - **isinf parity**: inf positions match
   - **signbit parity**: sign of every ±inf and every ±0.0 matches
4. Fails with a clear message showing position, eager value, and compiled value

The initial PR will enable the flag on three seed ops (`copysign`, `atan2`,
`reciprocal`) to prove the mechanism works end-to-end. Additional ops can be
opted in via follow-up PRs.

---

## What this does NOT include (intentional scope limit)

- Exception parity testing (eager raises, compiled succeeds silently) — separate issue/PR
- NaN payload comparison — out of scope, PyTorch does not guarantee NaN payload propagation
- Non-floating-point types — signed zeros and signed infinities are a floating-point concept

---

## cc

`@pytorch/pt2` for the compile-vs-eager aspect

cc @mruberry @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by raravind007**
Fixes #188680

`torch.allclose` treats +0.0 and -0.0 as equal (IEEE 754 defines them as
equal under `==`), so the default compile-vs-eager correctness check in
`test_ops.py` cannot detect a compiler pass that silently normalizes
signed zero (e.g. `-0.0 -> +0.0`). This adds an opt-in check, off by
default, for ops where that distinction is part of the observable
behavior.

New `OpInfo` field `check_sign_sensitive_special_values: bool = False`.
When set, `test_sign_sensitive_special_values` runs the op's
`reference_inputs` (falling back to `sample_inputs`) through eager and
`torch.compile`, and asserts:
- isnan mask parity
- isinf mask parity
- signbit parity, restricted to positions where both outputs are finite
  zero (infinity sign is already enforced by the existing allclose-based
  correctness check, so it isn't re-checked here)

Seed ops for this PR, chosen because signed-zero/signed-infinity behavior
is part of their documented contract rather than an implementation detail:
- `copysign` - sign copying is the entire purpose of the op
- `atan2` - IEEE 754-2008 Sec 9.2.1 specifies signed-zero results at
  boundary inputs
- `reciprocal` - sign of a zero input propagates to the sign of the
  resulting infinity

Scope intentionally kept small per discussion on #188680 - this is a
targeted guardrail for a handful of ops with observable special-value
semantics, not a general exact-correctness requirement.

cc @morrison-turnansky

### CODE REVIEW

**Comment by raravind007** [comment]
@morrison-turnansky Can I move PR from Draft to ready for review?

**Comment by morrison-turnansky** [comment]
yes

File: torch/testing/_internal/opinfo/core.py [hunk 1]
```diff
@@ -941,6 +941,23 @@ def __setattr__(self, name: str, value: Any) -> None:
 
     skip_correctness_check_compile_vs_eager: bool = False
 
+    # Whether to verify that isnan/isinf masks and signed-zero signbit are
```
Comment (morrison-turnansky) [inline torch/testing/_internal/opinfo/core.py]: can you make your comments less verbose?

File: test/test_ops.py [hunk 2]
```diff
@@ -132,6 +132,81 @@ def has_reduction_tag(op):
 # then they do not need to be compared to each other
 _ops_and_refs_with_no_numpy_ref = [op for op in ops_and_refs if op.ref is None]
 
+
+def _check_sign_sensitive(test_case, eager_out, compiled_out, *, label=""):
```
Comment (morrison-turnansky) [inline test/test_ops.py]: you are only using the function once inside another function and you have a underscore showing intent that this is a private function.  is there a reason not to move this within test_sign_sensitive_special_values, guarantee it is only used there, and consolidate a doc string and comment that are very similar

**Comment by raravind007** [comment]
Thanks for kicking off CI! Checked the results - everything is passing (nothing failed), so there's nothing to fix at the moment. I'll keep watching it and push a fix as soon as anything comes up.

On enabling it per run - since this PR is from a fork and I'm not a collaborator, GitHub requires a maintainer to manually approve the workflow each time new commits are pushed, so it won't rerun on its own. I'll ping you here when I push a fix so you can re-approve it.

**Comment by raravind007** [comment]
Good catch - moved `_check_sign_sensitive` inside `test_sign_sensitive_special_values` as a nested closure. It's now guaranteed single-use, no longer needs the `test_case` param (uses `self.fail` directly from the enclosing scope), and I consolidated the two overlapping docstrings/comments into one on the test method.

Pushed the update - since it's a new commit from a fork, could you re-approve the CI run when you get a chance?

File: test/test_ops.py [hunk 3]
```diff
@@ -132,6 +132,7 @@ def has_reduction_tag(op):
 # then they do not need to be compared to each other
 _ops_and_refs_with_no_numpy_ref = [op for op in ops_and_refs if op.ref is None]
 
+
```
Comment (morrison-turnansky) [inline test/test_ops.py]: nit, can you get rid of this whitespace change?

Comment (raravind007) [reply test/test_ops.py]: oops, can you check now?
